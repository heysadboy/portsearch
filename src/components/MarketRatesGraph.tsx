import { FC, useEffect, useState } from "react";
import { EStatusType, IGraphData, ILineData, IMarketRate, IPointData } from "../types";
import * as d3 from "d3";
import DateRange from "./DateRange";
import MarketPosition from "./MarketPosition";
import MultiLineChart from "./MultiLineChart";
import DataPointLabel from "./DataPointsLabel";
import Status from "./Status";

interface IMarketRatesGraphProps {
  data: IMarketRate[],
}

//Graph Data component to display information related to the graph
const MarketRatesGraph: FC<IMarketRatesGraphProps> = ({ data }) => {
  // Setting selected position average by default
  const [selectedPositions, setSelectedPositions] = useState<string[]>(["average"]);
  const [graphData, setGraphData] = useState<IGraphData>({});
  const [multiLineData, setMultiLineData] = useState<ILineData[]>([]);

  useEffect(() => {
    //Prepare data and separate informtion based on the market position
    const marketLow: ILineData = { name: "low", color: "#D50000", values: [] };
    const marketAvg: ILineData = { name: "average", color: "#AA00FF", values: [] }
    const marketHigh: ILineData = { name: "high", color: "#2962FF", values: [] };

    data.forEach((item) => {
      // Parsing date here to use correctly with d3
      const lowObj: IPointData = {
        day: d3.timeParse("%Y-%m-%d")(item.day),
        value: item.low
      }

      const avgObj: IPointData = {
        day: d3.timeParse("%Y-%m-%d")(item.day),
        value: item.mean
      }

      const highObj: IPointData = {
        day: d3.timeParse("%Y-%m-%d")(item.day),
        value: item.high
      }

      marketLow.values.push(lowObj);
      marketAvg.values.push(avgObj);
      marketHigh.values.push(highObj);
    });

    setGraphData({ "low": marketLow, "average": marketAvg, "high": marketHigh })
  }, [data]);

  useEffect(() => {
    const tempData: ILineData[] = [];
    //This is changed everytime market position is selected fro the MarketPosition array
    selectedPositions.forEach((position) => {
      tempData.push(graphData[position])
    });

    setMultiLineData(tempData);
  }, [graphData, selectedPositions]);

  //Setting up domain information. Doing this because we have unprocessed response with all grouped data here.
  //In graph we only send processed response
  const domain = {
    xMinValue: d3.min(data, d => d.day),
    xMaxValue: d3.max(data, d => d.day),
    yMinValue: d3.min(data, d => d.low),
    yMaxValue: d3.max(data, d => d.high)
  }

  return (
    <div id="graph-container">
      <DateRange startDate={domain.xMinValue} endDate={domain.xMaxValue} />
      <div className="ui stackable two column grid">
        <div className="sixteen wide tablet thirteen wide computer column">
          {selectedPositions.length > 0 ? <MultiLineChart data={multiLineData} domain={domain} /> : <Status status={EStatusType.no_position_selected} />}
        </div>
        <div className="four wide tablet three wide computer column">
          <DataPointLabel info={`${data.length} Data Points`} />
          <div className="ui segment market-position-container">
            <MarketPosition setSelectedPositions={setSelectedPositions} selectedPositions={selectedPositions} />
          </div>
        </div>
      </div>
    </div>
  )
}

export default MarketRatesGraph;