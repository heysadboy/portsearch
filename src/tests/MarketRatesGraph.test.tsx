import TestRenderer from "react-test-renderer";
import MarketRatesGraph from "../components/MarketRatesGraph";

const data = [
    {
        day: "2021-01-01",
        mean: 1615,
        low: 1037,
        high: 2436,
    },
    {
        day: "2021-01-02",
        mean: 1611,
        low: 1037,
        high: 2436,
    },
    {
        day: "2021-01-03",
        mean: 1613,
        low: 1007,
        high: 2436,
    }
]
test("MarketRatesGraph Component Testing.", () => {
    const component = TestRenderer.create(
        <MarketRatesGraph data={data} />
    );

    let tree: any = component.toJSON();
    expect(tree).toMatchSnapshot();
});