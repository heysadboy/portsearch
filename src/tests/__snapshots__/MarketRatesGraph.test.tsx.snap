// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarketRatesGraph Component Testing. 1`] = `
<div
  id="graph-container"
>
  <div
    className="ui basic label date-container"
  >
    <i
      className="calendar outline icon"
    />
    <span>
      2021-01-01
    </span>
    <i
      className="date-range-icon right arrow icon"
    />
    <span>
      2021-01-03
    </span>
  </div>
  <div
    className="ui stackable two column grid"
  >
    <div
      className="sixteen wide tablet thirteen wide computer column"
    >
      <div
        id="chart"
      />
    </div>
    <div
      className="four wide tablet three wide computer column"
    >
      <div
        className="ui orange label data-point-container"
      >
        3 Data Points
      </div>
      <div
        className="ui segment market-position-container"
      >
        <div
          className="grouped fields"
        >
          <div
            className="market-position-title"
          >
            Market Position
          </div>
          <div
            className="field"
          >
            <div
              className="ui checkbox market-checkbox"
            >
              <input
                name="high"
                onChange={[Function]}
                type="checkbox"
              />
              <label>
                High
              </label>
            </div>
          </div>
          <div
            className="field"
          >
            <div
              className="ui checkbox market-checkbox"
            >
              <input
                defaultChecked={true}
                name="average"
                onChange={[Function]}
                type="checkbox"
              />
              <label>
                Average
              </label>
            </div>
          </div>
          <div
            className="field"
          >
            <div
              className="ui checkbox market-checkbox"
            >
              <input
                name="low"
                onChange={[Function]}
                type="checkbox"
              />
              <label>
                Low
              </label>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
